File: codingInterviews/数字字符串转化成IP地址.java
import java.util.*;

/**
 * 题目描述
 * 现在有一个只包含数字的字符串，将该字符串转化成IP地址的形式，返回所有可能的情况。
 * 例如：
 * 给出的字符串为"25525522135",
 * 返回["255.255.22.135", "255.255.221.35"]. (顺序没有关系)
 * 示例1
 * 输入
 * "25525522135"
 * 返回值
 * ["255.255.22.135","255.255.221.35"]
 */
public class Solution {
    /**
     *
     * @param s string字符串
     * @return string字符串ArrayList
     */
    ArrayList<String> result = new ArrayList<>();
    public ArrayList<String> restoreIpAddresses (String s) {
        // write code here
        if (s.length()>12||s.length()<4){
            return result;
        }
        each(s,0,3);
        return result;
    }
    public void each (String s ,int idx,int count){
        if (count==0){
            String[] strs=s.split("\\.");
            if (strs.length==4){
                for(String str:strs){
                    if(str.length() > 1 && str.charAt(0) == '0') return;  //最前面的数字不能为0
                    if(Integer.valueOf(str) > 255) return;  //每一位都不能大于255
                }
                result.add(s);
                return;
            }
        }
        if (idx<s.length()-1){
            each(s.substring(0,idx+1)+'.'+s.substring(idx+1,s.length()),idx+2,count-1);
        }if (idx<s.length()-2){
            each(s.substring(0,idx+2)+'.'+s.substring(idx+2,s.length()),idx+3,count-1);
        }if (idx<s.length()-3){
            each(s.substring(0,idx+3)+'.'+s.substring(idx+3,s.length()),idx+4,count-1);
        }
    }
}

File: code-go/数字字符串转化成IP地址.go
package code_go

/**
题目描述
现在有一个只包含数字的字符串，将该字符串转化成IP地址的形式，返回所有可能的情况。
例如：
给出的字符串为"25525522135",
返回["255.255.22.135", "255.255.221.35"]. (顺序没有关系)
示例1
输入
"25525522135"
返回值
["255.255.22.135","255.255.221.35"]
*/
func restoreIpAddresses(s string) []string {
	var result []string
	backrack(s, "", 0, 0, &result)
	return result
}

func backrack(s string, tmp string, count, idx int, res *[]string) {
	if count == 3 {
		if check(s, idx, len(s)) {
			tmp += s[idx:]
			*res = append(*res, tmp)
		}
		return
	}
	for i := 1; i <= 3; i++ {
		if check(s, idx, idx+i) {
			backrack(s, tmp+s[idx:idx+i]+".", count+1, idx+i, res)
		}
	}
}

func check(bs string, i, j int) bool {
	if j-i > 3 || j-i <= 0 {
		return false
	}
	var v int
	for k := i; k < j; k++ {
		if k >= len(bs) {
			return false
		}
		if bs[k] < '0' || bs[k] > '9' {
			return false
		}
		if k == i+1 && v == 0 {
			return false
		}
		v = v*10 + int(bs[k]-'0')
		if v > 255 {
			return false
		}
	}
	return true
}
